### PR TITLE
Defer generating dereference of vft-symbol in JProfilingValue

### PR DESCRIPTION
**Pull request by hzongaro**
In the process of finding values to profile, the method `TR_JProfilingValue::cleanUpAndAddProfilingCandidates` would insert calls to the `jProfileValueWithNullCHK` non-helper for `instanceof` and `checkcast` operations, with the child of each call being an indirect load of the `<vft-symbol>` for the value that is to be profiled.

During `TR_JProfilingValue::addProfilingTrees`, those calls would be transformed to guard the dereference of the indirect load with a test of whether the value is a null reference.

However, if generation of profiling trees is halted early during `TR_JProfilingValue::lowerCalls`, any value beneath the `jProfileValueWithNullCHK` will simply be left anchored in its original position, unguarded.  This can result in a segmentation fault if the value that was to be profiled was a null reference.

This change delays introducing the indirect load of the `<vft-symbol>` until `TR_JProfilingValue::lowerCalls`, so that the indirect load is properly guarded.  If generation of profiling trees is halted early, the original value will be left behind anchored, allowing it be properly evaluated in its original relative position.

Fixes:  #21822

### CODE REVIEW

**Comment by hzongaro** [comment]
@r30shah, may I ask you to review this change?

File: runtime/compiler/optimizer/JProfilingValue.cpp [hunk 1]
```diff
@@ -606,8 +612,7 @@ TR_JProfilingValue::addProfilingTrees(
        * In case some one else decides to add VFT profiling around type test in other pass, it should also add the null test to guard
        * the profiling, as due to other optimizations it might be possible that VFT load is uncommoned and converted to load from the temp slot or register.
        */
-      TR_ASSERT_FATAL(value->getNumChildren() == 1, "JProfilingValue : NULL check can only be done when object node is available.");
-      TR::Node *nullTest = TR::Node::createif(TR::ifacmpeq, value->getFirstChild(), TR::Node::aconst(value, 0), mainlineReturn->getEntry());
+      TR::Node *nullTest = TR::Node::createif(TR::ifacmpeq, value, TR::Node::aconst(value, 0), mainlineReturn->getEntry());
```
Comment (r30shah) [inline runtime/compiler/optimizer/JProfilingValue.cpp:617]: We need to make sure that we create a VFT load from object after NULL test and use that value for the inlined quick test code for profiling value.

Comment (hzongaro) [reply runtime/compiler/optimizer/JProfilingValue.cpp:617]: I see - I had missed the [use of `value`](https://github.com/eclipse-openj9/openj9/blob/6bcf423253d53159f72b51d547782f210e0957c4/runtime/compiler/optimizer/JProfilingValue.cpp#L636) in `quickTestBlock`.

File: runtime/compiler/optimizer/JProfilingValue.cpp [hunk 2]
```diff
@@ -476,22 +482,22 @@ TR_JProfilingValue::lowerCalls()
...
- *          | istorei                      |                |                          |
- *          |    al/aiadd                  |                v                          |
- *          |       aconst <table address> |        |-------------------------------|  |
- *          |       l/imul                 |        | helper                        |  |
- *          |           l/iselect          |        |-------------------------------|  |
- *          |          width               |        | call TR_jProfile32/64BitValue |  |
- *          |     iadd                     |        |    value                      |  |
- *          |        iloadi                |        |    table address              |  |
- *          |           => al/aiadd        |        |-------------------------------|  |
- *          |        iconst 1              |                         |                 |
```
Comment (hzongaro) [inline runtime/compiler/optimizer/JProfilingValue.cpp:492]: I accidentally removed the `iconst 1` from this diagram.  I will fix that along with addressing other review comments.

File: runtime/compiler/optimizer/JProfilingValue.cpp (hunk 1 above)
Comment (hzongaro) [reply runtime/compiler/optimizer/JProfilingValue.cpp:617]: Fixed in commit 85223c14b211c916c29f333009a0b8efce7a00f7

File: runtime/compiler/optimizer/JProfilingValue.cpp (hunk 2 above)
Comment (hzongaro) [reply runtime/compiler/optimizer/JProfilingValue.cpp:492]: Fixed in commit 85223c14b211c916c29f333009a0b8efce7a00f7

**Comment by hzongaro** [comment]
@r30shah, I have addressed your initial review comment.  This is ready for another review.  Once you are OK with the changes, I will squash the commits.

**Review by r30shah — APPROVED** [review]
LGTM.

**Comment by r30shah** [comment]
jenkins test sanity all jdk8,jdk21

**Comment by r30shah** [comment]
All build and test jobs passes. Build test pipeline for 64-bit windows is set to false, though both downstream job (Build and Test) passes. I do not think this is real issue, change is good to merge, so going to merge.
